File: lib/wallpaper.rb
require 'net/http'
require 'json'
require 'fileutils'
require 'uri'
require 'thor'

class Wallpaper < Thor
  API_URL = URI("https://wallhaven.cc/api/v1/search?sorting=random&q=lowlight&atleast=1920x1080")
  SAVE_DIR = '/data/pictures/wallpaper/'
  FILENAMES = ['wallpaper-left.jpg', 'wallpaper-right.jpg']

  desc "download [SCREEN]", "Download wallpaper from Wallhaven. SCREEN can be 'left' or 'right'"
  option :apply, type: :boolean, default: false, desc: "Apply wallpaper to sway"
  option :mnt, type: :string, default: "", desc: "Mount point"
  def download(screen = nil)
    puts "Downloading wallpapers..."
    data = fetch_json
    urls = random_wallpapers(data)
    download_wallpaper(urls.first, FILENAMES.first) if screen == 'left' || screen.nil?
    download_wallpaper(urls.last, FILENAMES.last) if screen == 'right' || screen.nil?
    update_sway_wallpaper if options[:apply]
  end

  private

  def configure_ssl(http)
    http.use_ssl = true
    http.cert_store = OpenSSL::X509::Store.new
    http.cert_store.set_default_paths
    http.cert_store.flags = 0  # Disable CRL checking
  end

  def fetch_json
    http = Net::HTTP.new(API_URL.host, API_URL.port)
    configure_ssl(http)

    response = http.get(API_URL.request_uri)
    json = JSON.parse(response.body)
    data = json && json["data"]
    raise "No wallpaper found" unless data&.any?
    data
  end

  def random_wallpapers(data)
    puts "Taking a sample from #{data.size} images"
    urls = data.sample(2).map { URI(it['path']) }

    raise "Not enough wallpapers found" if urls.count < 2

    urls
  end

  def download_wallpaper(url, filename)
    http = Net::HTTP.new(url.host, url.port)
    configure_ssl(http)

    response = http.get(url.path)

    path = File.join(options[:mnt], SAVE_DIR, filename)
    File.open(path, 'wb') { |file| file.write(response.body) }

    puts "Saved wallpaper to #{path}"
  end

  def update_sway_wallpaper
    `swaymsg output eDP-1 background /data/pictures/wallpaper/wallpaper-left.jpg fill`
    `swaymsg output DP-2 background /data/pictures/wallpaper/wallpaper-left.jpg fill`
    `swaymsg output DP-3 background /data/pictures/wallpaper/wallpaper-right.jpg fill`
  end
end

File: src/wallpaper.rs
//! Equivalent of lib/wallpaper.rb: download random wallpapers from Wallhaven.

use anyhow::{bail, Context, Result};
use rand::seq::SliceRandom;
use std::io::Read;

use crate::system;

const API_URL: &str = "https://wallhaven.cc/api/v1/search?sorting=random&q=lowlight&atleast=1920x1080";
const SAVE_DIR: &str = "/data/pictures/wallpaper";
const FILENAMES: [&str; 2] = ["wallpaper-left.jpg", "wallpaper-right.jpg"];

#[derive(serde::Deserialize)]
struct ApiResponse {
    data: Vec<Image>,
}

#[derive(serde::Deserialize)]
struct Image {
    path: String,
}

pub fn download(screen: Option<&str>, apply: bool, mnt: &str) -> Result<()> {
    println!("Downloading wallpapers...");
    let images = fetch_images()?;
    let urls = pick_two(&images)?;

    if matches!(screen, None | Some("left")) {
        save(&urls[0], FILENAMES[0], mnt)?;
    }
    if matches!(screen, None | Some("right")) {
        save(&urls[1], FILENAMES[1], mnt)?;
    }
    if apply {
        update_sway()?;
    }
    Ok(())
}

fn fetch_images() -> Result<Vec<Image>> {
    let body: ApiResponse = ureq::get(API_URL)
        .call()
        .context("fetching wallhaven API")?
        .into_json()
        .context("parsing wallhaven response")?;
    if body.data.is_empty() {
        bail!("No wallpaper found");
    }
    Ok(body.data)
}

fn pick_two(images: &[Image]) -> Result<Vec<String>> {
    println!("Taking a sample from {} images", images.len());
    let mut rng = rand::thread_rng();
    let picks: Vec<&Image> = images.choose_multiple(&mut rng, 2).collect();
    if picks.len() < 2 {
        bail!("Not enough wallpapers found");
    }
    Ok(picks.into_iter().map(|i| i.path.clone()).collect())
}

fn save(url: &str, filename: &str, mnt: &str) -> Result<()> {
    let resp = ureq::get(url).call().context("fetching wallpaper image")?;
    let mut bytes = Vec::new();
    resp.into_reader()
        .read_to_end(&mut bytes)
        .context("reading wallpaper body")?;

    // Naive concatenation matches Ruby File.join semantics: an empty mnt
    // leaves SAVE_DIR's leading slash intact; a non-empty mnt is prepended
    // verbatim (e.g. "/mnt" + "/data/..." = "/mnt/data/...").
    let path = format!("{mnt}{SAVE_DIR}/{filename}");
    std::fs::write(&path, &bytes).with_context(|| format!("writing {path}"))?;
    println!("Saved wallpaper to {path}");
    Ok(())
}

fn update_sway() -> Result<()> {
    let flag = "/tmp/sway-wallpaper-changing";
    std::fs::write(flag, b"")?;
    for (output, file) in [
        ("eDP-1", FILENAMES[0]),
        ("DP-2", FILENAMES[0]),
        ("DP-3", FILENAMES[1]),
    ] {
        let path = format!("{SAVE_DIR}/{file}");
        system::run_capture_quiet(&format!("swaymsg output {output} background {path} fill"))?;
    }
    std::thread::sleep(std::time::Duration::from_secs(1));
    let _ = std::fs::remove_file(flag);
    Ok(())
}
